scintillometry/psrio/psrfits_io.py

```diff
@@ -0,0 +1,383 @@
+
+    @property
+    def nbin(self):
+        return int(self.header['NBIN'])
+
+    @property
+    def samples_per_frame(self):
+        try:
+            return int(self.header['NSBLK'])
+        except:
+            return int(np.prod(self.data_shape)/(self.nrow * self.nchan *
+                                                 self.npol * self.nbin))
+
+    @samples_per_frame.setter
+    def samples_per_frame(self, val):
+        self.header['NSBLK'] = val
+
+    @property
+    def shape(self):
+        raw_shape = self.raw_shape
+        new_shape = namedtuple('shape', ['nsample','nbin', 'nchan', 'npol'])
+        result = new_shape(raw_shape.nrow * raw_shape.samples_per_frame,
+                           raw_shape.nbin, raw_shape.nchan, raw_shape.npol)
+        return result
+
+    @property
+    def raw_shape(self):
+        r_shape = namedtuple('shape', ['nrow', 'samples_per_frame','nbin',
+                                       'nchan','npol'])
+        result = r_shape(self.nrow, self.samples_per_frame, self.nbin,
+                         self.nchan, self.npol)
+        return result
+
+    @property
+    def data_shape(self):
+        # Data are save in the fortran order. Reversed from the header label.
+        d_shape_raw = self.data['DATA'].shape
+        if self.mode == "SEARCH":
+            d_shape_header = (self.nbin, self.nchan, self.npol,
+                              self.samples_per_frame)
+        else:
+            d_shape_header = (self.nbin, self.nchan, self.npol)
+        if d_shape_raw != (self.nrow, ) + d_shape_header[::-1]:
+            raise ValueError("Data shape does not match with the header"
+                             " information")
+        if self.mode == "SEARCH":
+            d_shape = namedtuple('d_shape', ['nrow', 'nsample', 'npol', 'nchan',
+                                             'nbin'])
+            result  = d_shape(self.nrow,  self.samples_per_frame, self.npol,
+                              self.nchan, self.nbin)
+        else:
+            d_shape = namedtuple('d_shape', ['nrow', 'npol', 'nchan', 'nbin'])
+            result  = d_shape(self.nrow,  self.npol, self.nchan, self.nbin)
+        return result
+
+    @property
+    def polarization(self):
+        return self.header['POL_TYPE']
+
+    @polarization.setter
+    def polarization(self, val):
+        self.header['POL_TYPE'] = val
+
+    @property
+    def frequency(self):
+        if 'DAT_FREQ' in self.columns.names:
+            freqs = u.Quantity(self.data['DAT_FREQ'],
+                               u.MHz, copy=False)
+            assert np.isclose(freqs, freqs[0],
+                              atol=np.finfo(float).eps * u.MHz).all(), \
+                "Frequencies are different within one subint rows."
+            freqs = freqs[0]
+        else:
+            # Get the frequency from the header HDU
+            try:
+                freqs = self.header_hdu.frequency
+            except:
+                freqs = None
```

Comment:
Don't return `None`, but rather raise an `AttributeError`?

Reply:
Some profile PSRFITS is not so nice to provide the frequency information. Maybe we can give a warning?